### ISSUE TITLE
MSBuild task failing with: MSB4044: The "GetMinimumOSVersion" task was not given a value for the required parameter "SdkVersion"

### ISSUE THREAD

**Issue by bijington** [xamarin/xamarin-macios#10109]
### Steps to Reproduce

1. Run this command `"C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Current\Bin\msbuild" MySolution.sln /p:Configuration="AppStore" /p:Platform="iPhone" /p:ServerAddress="" /p:ServerUser=""`

### Expected Behavior

It continues to successfully compile.

### Actual Behavior

It fails with error message:

> [GetMinimumOSVersion] C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.Shared.targets(230, 3): error MSB4044: The "GetMinimumOSVersion" task was not given a value for the required parameter "SdkVersion".

It was previously working with Visual Studio 16.7.7 but I don't recall the Xamarin.iOS version sadly.

### Additional detail

Upon digging I believe I have traced the source of the error back to these lines not being set in Xamarin.Shared.props.

```
<!-- SdkVersion -->
<_SdkVersion Condition="'$(_PlatformName)' == 'macOS'">$(MacOSXSdkVersion)</_SdkVersion>
<_SdkVersion Condition="'$(_PlatformName)' != 'macOS'">$(MtouchSdkVersion)</_SdkVersion>
```

I haven't been able to find out what sets MtouchSdkVersion and therefore why it isn't being set. A comment from just above the quoted section above states:

> Sometimes we've used different variable names for the same thing for Xamarin.iOS and Xamarin.Mac projects. Here we try to unify those variables

Could it be that this unification has changed the behaviour of something that used to work?

For the record I have also raised this issue on [Stack Overflow](https://stackoverflow.com/questions/64876303/visual-studio-update-16-8-1-lead-to-ci-build-failures) and will update here accordingly if I get to an answer sooner.

### Environment

```
Microsoft Visual Studio Professional 2019
Version 16.8.1
VisualStudio.16.Release/16.8.1+30711.63
Microsoft .NET Framework
Version 4.8.04084

Installed Version: Professional

Visual C++ 2019   00435-60000-00000-AA092
Microsoft Visual C++ 2019

ASP.NET and Web Tools 2019   16.8.550.19892
ASP.NET and Web Tools 2019

ASP.NET Core Razor Language Services   16.1.0.2052803+84e121f1403378489b842e1797df2f3f5a49ac3c
Provides languages services for ASP.NET Core Razor.

ASP.NET Web Frameworks and Tools 2019   16.8.550.19892
For additional information, visit https://www.asp.net/

Azure App Service Tools v3.0.0   16.8.550.19892
Azure App Service Tools v3.0.0

Azure Functions and Web Jobs Tools   16.8.550.19892
Azure Functions and Web Jobs Tools

C# Tools   3.8.0-5.20519.18+4c195c3ac1974edcefa76774d7a59a2350ec55fa
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Extensibility Message Bus   1.2.6 (master@34d6af2)
Provides common messaging-based MEF services for loosely coupled Visual Studio extension components communication and integration.

IntelliCode Extension   1.0
IntelliCode Visual Studio Extension Detailed Info

Microsoft Azure Tools   2.9
Microsoft Azure Tools for Microsoft Visual Studio 2019 - v2.9.30924.1

Microsoft Continuous Delivery Tools for Visual Studio   0.4
Simplifying the configuration of Azure DevOps pipelines from within the Visual Studio IDE.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft Library Manager   2.1.113+g422d40002e.RR
Install client-side libraries easily to any web project

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual C++ Wizards   1.0
Microsoft Visual C++ Wizards

Microsoft Visual Studio Tools for Containers   1.1
Develop, run, validate your ASP.NET Core applications in the target environment. F5 your application directly into a container with debugging, or CTRL + F5 to edit & refresh your app without having to rebuild the container.

Microsoft Visual Studio VC Package   1.0
Microsoft Visual Studio VC Package

Mono Debugging for Visual Studio   16.8.43 (00471f8)
Support for debugging Mono processes with Visual Studio.

NuGet Package Manager   5.8.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/

ProjectServicesPackage Extension   1.0
ProjectServicesPackage Visual Studio Extension Detailed Info

SQL Server Data Tools   16.0.62010.06180
Microsoft SQL Server Data Tools

TypeScript Tools   16.0.21016.2001
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   3.8.0-5.20519.18+4c195c3ac1974edcefa76774d7a59a2350ec55fa
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual F# Tools   16.8.0-beta.20507.4+da6be68280c89131cdba2045525b80890401defd
Microsoft Visual F# Tools

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

Visual Studio Container Tools Extensions   1.0
View, manage, and diagnose containers within Visual Studio.

Visual Studio Tools for Containers   1.0
Visual Studio Tools for Containers

VisualStudio.DeviceLog   1.0
Information about my package

VisualStudio.Foo   1.0
Information about my package

VisualStudio.Mac   1.0
Mac Extension for Visual Studio

Xamarin   16.8.000.255 (d16-8@d002176)
Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.

Xamarin Designer   16.8.0.507 (remotes/origin/d16-8@e87b24884)
Visual Studio extension to enable Xamarin Designer tools in Visual Studio.

Xamarin Templates   16.8.112 (86385a3)
Templates for building iOS, Android, and Windows apps with Xamarin and Xamarin.Forms.

Xamarin.Android SDK   11.1.0.17 (d16-8/c0e2b8e)
Xamarin.Android Reference Assemblies and MSBuild support.
    Mono: be2226b
    Java.Interop: xamarin/java.interop/d16-8@79d9533
    ProGuard: Guardsquare/proguard/proguard6.2.2@ebe9000
    SQLite: xamarin/sqlite/3.32.1@1a3276b
    Xamarin.Android Tools: xamarin/xamarin-android-tools/d16-8@2fb1cbc


Xamarin.iOS and Xamarin.Mac SDK   14.4.1.3 (e30c41de3)
Xamarin.iOS and Xamarin.Mac Reference Assemblies and MSBuild support.
```

### Build Logs

This is a build log from our TeamCity build agent running the command mentioned in **Steps to reproduce**:

[XenplateDC.iOS\XenplateDC.iOS.csproj] _SayHello (14s)
[22:06:26][_SayHello] Executing SayHello Task to establish a connection to a Mac Server. 
						 Properties: 
							SessionId=blanked, 
							ServerPort=, 
							ServerAddress=blanked, 
							ServerUser=blanked, 
							ServerPassword=, 
							SSHKey=, 
							SSHPassPhrase=,
							AppName=XenplateDC.iOS,
							VisualStudioProcessId=,
							ContinueOnDisconnected=true
[22:06:26][_SayHello] SayHello (14s)
[22:06:26][SayHello] Connection Information for App 'XenplateDC.iOS'. Address: blanked, Port: 0, User: blanked
[22:06:26][SayHello] Connecting to Mac server blanked...
[22:06:27][SayHello] [xma]: Checking Broker 16.8.000.255 installation...
[22:06:31][SayHello] [xma]: Starting Broker 16.8.000.255 in port 55120...
[22:06:31][SayHello] C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.Messaging.targets(59, 3): [xma]: Failed to execute 'which mono64': ExitStatus = 1
[22:06:31][SayHello] [xma]: A valid Mono installation has been found at '/Library/Frameworks/Mono.framework/Versions/Current/bin/mono64'
[22:06:31][SayHello] [xma]: Uploading 'app.plist' (1/1 KB) 100.00%...
[22:06:32][SayHello] [xma]: Broker 16.8.000.255 started successfully
[22:06:34][SayHello] [xma]: Agent Broker 16.8.000.255 is running
[22:06:34][SayHello] [xma]: Checking Build 16.8.000.255 installation...
[22:06:39][SayHello] [xma]: build4f69ce9d2b129c07504c6023b7d5609e9f80626e32834520f2506564bcfa16ac2088Shaun received ping. Sending acknowledge...
[22:06:40][SayHello] [xma]: Starting Agent Build 16.8.000.255...
[22:06:40][SayHello] [xma]: Starting Build 16.8.000.255 in port 55120...
[22:06:40][SayHello] [xma]: Uploading 'app.plist' (1/1 KB) 100.00%...
[22:06:40][SayHello] [xma]: Agent Build 16.8.000.255 is running
[22:06:41][SayHello] [xma]: Build 16.8.000.255 started successfully
[22:06:41][SayHello] [xma]: Agent Build 16.8.000.255 is running
[22:06:41][XenplateDC.iOS\XenplateDC.iOS.csproj] _VerifyBuildSignature (2s)
[22:06:43][XenplateDC.iOS\XenplateDC.iOS.csproj] PrepareForBuild
[22:06:43][PrepareForBuild] MakeDir
[22:06:43][XenplateDC.iOS\XenplateDC.iOS.csproj] _GetMinimumOSVersion
[22:06:43][_GetMinimumOSVersion] GetMinimumOSVersion
[22:06:43][GetMinimumOSVersion] C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.Shared.targets(230, 3): error MSB4044: The "GetMinimumOSVersion" task was not given a value for the required parameter "SdkVersion".

### Example Project (If Possible)

> VS bug [#1248227](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1248227)

**Comment by binesh1991** [comment]
This has broken our CI builds as well. Performing a rebuild within Visual Studio 2019 for the same configuration works fine however.
The msbuild command I've tried was
`"C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Current\Bin\MSBuild.exe" "<path to sln file>" /t:Rebuild "/p:Configuration=Ad-Hoc" "/p:Platform=iPhone" /p:ServerAddress=<ip> /p:ServerUser=<username> /p:ServerPassword=<password> /nr:false /v:m /p:RestorePackages=false /p:IpaPackageDir="<output path>"`

**Comment by rolfbjarne** [comment]
Could you attach a diagnostic build log? The SDK version should default to the SDK version shipped with the Xcode you're using, and for some reason that doesn't seem to happen.

**Comment by binesh1991** [comment]
[msbuild_output.txt](https://github.com/xamarin/xamarin-macios/files/5558641/msbuild_output.txt)
Here's the log of msbuild command with verbosity set to diagnostic level

**Comment by bijington** [comment]
@binesh1991 Thanks you beat me to it

**Comment by rolfbjarne** [comment]
The DetectSdkLocations task is where we find the default SDK version, but it seems there is a problem with the connection:

```
Task "DetectSdkLocations" (TaskId:44)
  Task Parameter:SessionId=66c26c9cb20f009947a0e92e7e7df6f56b7b343165c2b3680d5554b422917028 (TaskId:44)
  Task Parameter:TargetFrameworkMoniker=Xamarin.iOS,Version=v1.0 (TaskId:44)
  Task Parameter:TargetArchitectures=ARM64 (TaskId:44)
  DetectSdkLocations: 2020-11-18T13:27:56.2549670+05:30 - Started (TaskId:44)
  DetectSdkLocations: 2020-11-18T13:27:56.2549670+05:30 - Initializing (TaskId:44)
  [xma]: Trying to get a Build Connection for Session '66c26c9cb20f009947a0e92e7e7df6f56b7b343165c2b3680d5554b422917028': Xamarin.VisualStudio.Build.BuildConnection.66c26c9cb20f009947a0e92e7e7df6f56b7b343165c2b3680d5554b422917028, Lifetime: Build (TaskId:44)
  [xma]: The Build Connection for Session '66c26c9cb20f009947a0e92e7e7df6f56b7b343165c2b3680d5554b422917028' is not active and needs to be connected with the proper host and user information (TaskId:44)
  DetectSdkLocations: 2020-11-18T13:27:56.2639432+05:30 - Failed to initialize (TaskId:44)
  Output Property: _SdkIsSimulator=False (TaskId:44)
  Output Property: _IsXcode8=False (TaskId:44)
Done executing task "DetectSdkLocations". (TaskId:44)
```

@emaf can you have a look?

**Comment by binesh1991** [comment]
To add some more info, I have recreated this issue on two Mac build hosts running Xcode 12.2 and 12.0.1. Both Macs have latest VS 2019 for Mac installed.

**Comment by bijington** [comment]
I can confirm that I have the same entry in my log aside from different session ids obviously. Plus I am also running Xcode 12.2 and the latest VS2019 for Mac.

**Comment by emaf** [comment]
Hi all, I'm sorry about this problem, we're investigating it, but it looks like the build process is getting disconnected after running the clean targets, so if you run Build instead of Rebuild you shouldn't face this. 

In the meantime, a potential workaround to unblock Rebuilds would be to edit `<VS Install path>\MSBuild\Xamarin\iOS\Xamarin.iOS.Windows.After.targets` and remove the `_SayGoodbye` entry from the `CleanDependsOn` property group.

Please let us know if that unblocks your CI builds.

**Comment by binesh1991** [comment]
Thank you @emaf for the workaround. I'll stick with editing the Xamarin.iOS.Windows.After.targets file as that's the quickest option for me. I tried a new CI build (using Jenkins) after doing that but looks like msbuild is hanging after completion. Due to this, I had to terminate it manually from Task Manager for the rest of the build process to continue. So there's still some issue (probably [this](https://developercommunity2.visualstudio.com/t/Using-msbuildexe-from-the-command-line-/1193752) again).

Also @bijington, I think you're just doing a build (not rebuild) according to the command given? Anyway, it's better to post this workaround on Stack Overflow as well.

**Comment by bijington** [comment]
@emaf thank you for the speedy response! I can confirm that your workaround of modifying the targets file has worked for me.

@binesh1991 yes I will update Stack Overflow to cover the workaround for now.

**Comment by emaf** [comment]
Thanks for confirming the workaround worked! 

@binesh1991 could you share a diagnostic build output after the build stops responding? The workaround shouldn't make that happen, or at least for a rebuild so we'd like to understand what's going on there.

**Comment by binesh1991** [comment]
I think that happened only for the first one after editing the file. The next two builds went through fine, so we can live with the workaround for the time being. Thanks!

**Comment by stesvis** [comment]
> Hi all, I'm sorry about this problem, we're investigating it, but it looks like the build process is getting disconnected after running the clean targets, so if you run Build instead of Rebuild you shouldn't face this.
> 
> In the meantime, a potential workaround to unblock Rebuilds would be to edit `<VS Install path>\MSBuild\Xamarin\iOS\Xamarin.iOS.Windows.After.targets` and remove the `_SayGoodbye` entry from the `CleanDependsOn` property group.
> 
> Please let us know if that unblocks your CI builds.

@emaf How do you do this on mac?

**Comment by bijington** [comment]
> > Hi all, I'm sorry about this problem, we're investigating it, but it looks like the build process is getting disconnected after running the clean targets, so if you run Build instead of Rebuild you shouldn't face this.
> 
> > 
> 
> > In the meantime, a potential workaround to unblock Rebuilds would be to edit `<VS Install path>\MSBuild\Xamarin\iOS\Xamarin.iOS.Windows.After.targets` and remove the `_SayGoodbye` entry from the `CleanDependsOn` property group.
> 
> > 
> 
> > Please let us know if that unblocks your CI builds.
> 
> 
> 
> @emaf How do you do this on mac?

As far as I understand the problem it is only an issue when connecting from a Windows machine to a Mac. Therefore you only need to change the Xamarin file on your **Windows** machine

**Comment by emaf** [comment]
@stesvis as @bijington said this problem is only affecting Xamarin iOS builds on Windows, so there's no change needed on the Mac. 

If you're facing the same error reported here from VS for Mac I'd suggest you report it as a separate issue because this problem is caused by a target that only runs on Windows so you may be facing a different problem.

Thanks!

**Comment by stesvis** [comment]
@emaf and @bijington yes i have the same problem on Mac. I will open a new issue.

**Comment by mclillill** [comment]
@xamarin-release-manager where can i find the new issue where this problem is beeing tracked?

**Comment by bijington** [comment]
@mclillill I am hoping this just means a new release with it fixed will be coming very soon

**Comment by stesvis** [comment]
In my case on the Mac the problem was a corrupted Info.plist that I didn't notice (due to a merge).
Not sure why the error message was so off tho.

**Comment by mauroa** [comment]
The issue on Windows has been fixed and will be available in the next 16.8 Service Release and 16.9 Preview. Thanks

**Comment by jgold6** [comment]
Is it likely that these two reported issues will also be resolved with this fix?
https://developercommunity2.visualstudio.com/t/Error-MSB4044:-The-CheckSwiftVersionsTa/1252463?preview=true
https://developercommunity2.visualstudio.com/t/CheckSwiftVersionTask/1252846?preview=true

**Comment by binesh1991** [comment]
Looks like it's back in 16.8.3 and a developer has found a workaround that's posted at https://stackoverflow.com/a/65207950/2492114

**Comment by lauxjpn** [comment]
> Looks like it's back in 16.8.3 and a developer has found a workaround that's posted at https://stackoverflow.com/a/65207950/2492114

I just hit this too with `16.8.3`.

This issue needs to be **reopened**.

**Comment by Delphinidae84** [comment]
I updated to 16.8.4 yesterday and I am still having the same issue. I agree with @lauxjpn the issue is still valid.

**Comment by kcrg** [comment]
Here we go again, updating the build tools is not going to be smooth, even after some bugfix releases. I also ran into this problem

**Comment by spouliot** [comment]
c.c. @mauroa @emaf

**Comment by maratoss** [comment]
It works for me
https://stackoverflow.com/questions/64876303/visual-studio-update-16-8-1-lead-to-ci-build-failures/65207950#65207950

**Comment by sheiksyedm** [comment]
I'm still having this issue in VS 16.8.4. @emaf  your suggested workaround for removing line in the CleanDependsOn property group not there in this version. Please check the image below 
![Xamarin.iOS.Windows.After.targets](https://user-images.githubusercontent.com/23059975/108503969-3a82ff80-72db-11eb-84ed-fe146263a080.png)

**Comment by mauroa** [comment]
Hi @sheiksyedm, could you please provide the binary logs with that new version so we can see what else could be causing this issue in your case?
You can get the logs following these steps:

1. Install the [Project System Tools](https://marketplace.visualstudio.com/items?itemName=VisualStudioProductTeam.ProjectSystemTools) extension
2. In Visual Studio, choose the View > Other Windows > Build Logging menu item.
3. Click on the "play" button.
4. Reproduce the issue
5. Select all entries in the build logging tool window
6. Right-click them, Save Logs
7. Attach the resulting zip file

Thanks in advanced

**Comment by lauxjpn** [comment]
> This solution works for me (16.8.4)
> https://stackoverflow.com/questions/64876303/visual-studio-update-16-8-1-lead-to-ci-build-failures/65207950#65207950

@sheiksyedm As @maratoss just referenced in the comment above yours, you should be able to do the following:

> This has reared it's head again in Visual Studio 2019 version 16.8.3. Looks like they've moved the problem to:
`<VS Install path>\MSBuild\Xamarin\iOS\Xamarin.Messaging.targets`
>
> I was able to get my builds running again by removing the `_SayGoodbye` dependency on `line 52` for the `_DisconnectAfterClean` target.
>
> `<Target Name="_DisconnectAfterClean" Condition="'$(IsRebuild)' != 'true'" DependsOnTargets="" />`

**Comment by nschoenberg** [comment]
Hi,
our Jenkins is using 16.8.5 and showing this issue. Tryed to use 16.8.6 to solve the issue without success. I used the workarround @lauxjpn quoted to solve the issue:

> This has reared it's head again in Visual Studio 2019 version 16.8.3. Looks like they've moved the problem to:
> <VS Install path>\MSBuild\Xamarin\iOS\Xamarin.Messaging.targets
> 
> I was able to get my builds running again by removing the _SayGoodbye dependency on line 52 for the _DisconnectAfterClean target.
> 
> \<Target Name="_DisconnectAfterClean" Condition="'$(IsRebuild)' != 'true'" DependsOnTargets="" />

For refernce find my diagnostic build output:
[diagnostic_build.txt](https://github.com/xamarin/xamarin-macios/files/6060016/diagnostic_build.txt)

**Comment by IlyaDhtvzGhbltn** [comment]
I'm using 16.8.6 and still have this problem.

**Comment by RC241** [comment]
Just updated VS to 16.9. Updated the Mac, paired the VS project to the Mac ok
Still having this problem, 
Microsoft Visual Studio\2019\Enterprise\MSBuild\Xamarin\iOS\Xamarin.Shared.targets(767,3): error MSB4044: The "GetMinimumOSVersion" task was not given a value for the required parameter "SdkVersion". [...\build\99be3a9f31a24401b36c2cb9117993dd\source\ios.csproj]

Trying the "targets" workaround will bypass this issue but failes to return the build result in the end and leaves our build-process hanging on the server.
We're going to set up some specific local environment to try to find out what the issue is.

**Comment by YZahringer** [comment]
I confirm, the issue persists with `Visual Studio for Windows 16.9.0` paired with `Visual Studio for Mac 8.9`

**Comment by YZahringer** [comment]
This has been blocking our continuous integration for 4 months. We were hoping for a fix with 16.9.0, but the problem persists. 

The previously mentioned workarounds do not work in 16.9.0 anymore. Any news?

**Comment by mauroa** [comment]
Hi everyone,
This issue has been fixed for good and will be available on the next VS 16.9 Service Release and the next VS 16.10 Preview 2. However, if you are working on a 16.9 installation, I'm attaching a file that you need to override in the VS installation folder as a workaround to get unblocked until we release the fixed version. So, to apply the workaround please:

- Go to: `<VS Install path>\MSBuild\Xamarin\iOS`
- Unzip the attached file (`Xamarin.Messaging.Build.targets.zip`) and paste the .targets file in the above folder. That should replace the existing file with the new one to fix the issue
- Try to reproduce the issue. It should work now

Please let me know if that unblocks you.
Thanks.

[Xamarin.Messaging.Build.targets.zip](https://github.com/xamarin/xamarin-macios/files/6126069/Xamarin.Messaging.Build.targets.zip)

**Comment by YZahringer** [comment]
@mauroa Thanks for the workaround. I tried it, I have another error now:
````
  Project "C:\*****.sln" (1) is building "C:\*****.csproj" (5:6) on node 1 (default targets).
  _GenerateBuildSessionId:
    Generated session id: *****
    Generated build app name: *****
  _SayHello:
    Executing SayHello Task to establish a connection to a Remote Server.
          Properties:
            SessionId=*****,
            ServerPort=,
            ServerAddress=*****,
            ServerUser=*****,
            ServerPassword=*******,
            SSHKey=,
            SSHPassPhrase=,
            AppName=*****,
            VisualStudioProcessId=,
            ContinueOnDisconnected=true
  C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.Messaging.Build.targets(122,7): error MSB4064: The "AgentsDirectory" parameter is not supported by the "SayHello" task. Verify the parameter exists on the task, and it is a settable public instance property. [C:\*****.csproj]
  C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.Messaging.Build.targets(107,3): error MSB4063: The "SayHello" task could not be initialized with its input parameters.  [C:\*****.csproj]
````

**Comment by mauroa** [comment]
Sorry about that @YZahringer. I've updated the attachment, please try again.
Thanks

**Comment by YZahringer** [comment]
@mauroa Thanks for the support. Now I have errors at `_CompileToNative` step:
````
  (_CompileToNative target) ->
  ...
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(425,3): warning : [xma]: Unable to authenticate using the following authentication method: KeyboardInteractiveAuthenticationMethod [C:\*****.csproj]
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(425,3): warning : Session operation has timed out [C:\*****.csproj]
  C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(425,3): warning :    at Renci.SshNet.Session.WaitOnHandle(WaitHandle waitHandle, TimeSpan timeout) [C:\*****.csproj]
  C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(425,3): warning :    at Renci.SshNet.Session.WaitOnHandle(WaitHandle waitHandle) [C:\*****.csproj]
  C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(425,3): warning :    at Renci.SshNet.KeyboardInteractiveAuthenticationMethod.Authenticate(Session session) [C:\*****.csproj]
  C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(425,3): warning :    at Xamarin.Messaging.Ssh.MessagingAuthenticationMethod.TryAuthenticate(Session session, AuthenticationMethod method) in C:\A\1\230\s\src\Xamarin.Messaging.Ssh\MessagingAuthenticationMethod.cs:line 118 [C:\*****.csproj]
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(425,3): warning G3E9910A3: auto property synthesis will not synthesize property 'inputImage' declared in protocol 'CIPhotoEffect' [-Wobjc-protocol-property-synthesis] [C:\*****.csproj]
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(425,3): warning G3E9910A3: auto property synthesis will not synthesize property 'inputImage' declared in protocol 'CISepiaTone' [-Wobjc-protocol-property-synthesis] [C:\*****.csproj]
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(425,3): warning G3E9910A3: auto property synthesis will not synthesize property 'intensity' declared in protocol 'CISepiaTone' [-Wobjc-protocol-property-synthesis] [C:\*****.csproj]
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(425,3): warning MT5215: References to 'System' might require additional -framework=XXX or -lXXX instructions to the native linker [C:\*****.csproj]
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(425,3): warning MT5215: References to 'System.Net.Security' might require additional -framework=XXX or -lXXX instructions to the native linker [C:\*****.csproj]
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(425,3): warning MT5215: References to 'kernel32' might require additional -framework=XXX or -lXXX instructions to the native linker [C:\*****.csproj]
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(425,3): warning MT5215: References to 'advapi32' might require additional -framework=XXX or -lXXX instructions to the native linker [C:\*****.csproj]
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(425,3): warning MT5215: References to 'kernel32' might require additional -framework=XXX or -lXXX instructions to the native linker [C:\*****.csproj]
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(425,3): warning MT5215: References to 'ntdll' might require additional -framework=XXX or -lXXX instructions to the native linker [C:\*****.csproj]
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(425,3): warning MT5215: References to 'bcrypt' might require additional -framework=XXX or -lXXX instructions to the native linker [C:\*****.csproj]
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(425,3): warning MT5215: References to 'crypt32' might require additional -framework=XXX or -lXXX instructions to the native linker [C:\*****.csproj]
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(425,3): warning MT5215: References to 'ncrypt' might require additional -framework=XXX or -lXXX instructions to the native linker [C:\*****.csproj]
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(425,3): warning MT5215: References to 'kernel32' might require additional -framework=XXX or -lXXX instructions to the native linker [C:\*****.csproj]

  "C:\*****.sln" (Restore;***** target) (1) ->
  "C:\*****.csproj" (default target) (5:6) ->
  (_CompileToNative target) ->
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(425,3): error MT5210: Native linking failed, undefined symbol: _FormatMessageW. Please verify that all the necessary frameworks have been referenced and native libraries are properly linked in. [C:\*****S.csproj]
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(425,3): error MT5210: Native linking failed, undefined symbol: _RegCloseKey. Please verify that all the necessary frameworks have been referenced and native libraries are properly linked in. [C:\*****.csproj]
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(425,3): error MT5210: Native linking failed, undefined symbol: _RegEnumValueW. Please verify that all the necessary frameworks have been referenced and native libraries are properly linked in. [C:\*****.csproj]
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(425,3): error MT5210: Native linking failed, undefined symbol: _RegOpenKeyExW. Please verify that all the necessary frameworks have been referenced and native libraries are properly linked in. [C:\*****.csproj]
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(425,3): error MT5210: Native linking failed, undefined symbol: _RegQueryInfoKeyW. Please verify that all the necessary frameworks have been referenced and native libraries are properly linked in. [C:\*****.csproj]
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(425,3): error MT5210: Native linking failed, undefined symbol: _RegQueryValueExW. Please verify that all the necessary frameworks have been referenced and native libraries are properly linked in. [C:\*****.csproj]
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(425,3): error MT5202: Native linking failed. Please review the build log. [C:\*****.csproj]
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(425,3): error : linker command failed with exit code 1 (use -v to see invocation) [C:\*****.csproj]
  
      64 Warning(s)
      8 Error(s)
````

**Comment by narciszait** [comment]
@mauroa with Visual Studio for Mac 8.9.x (i don't remember the number), we are seeing the same 6 Native linking failed errors
This is not happening on 8.8.10

**Comment by mauroa** [comment]
Thanks @YZahringer and @narciszait,
The native linking error is definitely another issue not related to this bug (and the originally reported root cause). As mentioned above, the original issue was fixed and will be available on the next 16.9 SR and 16.10 Preview.
Regarding the linker error, I suggest you to report it from scratch on DevCom in order to redirect and tackle it properly.

Thanks for your help and feedback.
I'm closing this issue now.

**Comment by Delphinidae84** [comment]
The issue is still persisting in the current 16.9.2 Release. Breaking our CI Pipeline for months now 😢  @mauroa

**Comment by mauroa** [comment]
@Delphinidae84 please attach the Xamarin.Messaging.Build.targets file from `<VS Install path>\MSBuild\Xamarin\iOS`

**Comment by taenito** [comment]
This is the file for the 16.9.2 Release.

This break our CI too.

[Xamarin.Messaging.Build.targets.zip](https://github.com/xamarin/xamarin-macios/files/6182580/Xamarin.Messaging.Build.targets.zip)

**Comment by mauroa** [comment]
I've verified the file that you attached and I can confirm that the fix is not in there, so it didn't make it for SR 2. I'll follow up on this and ensure it's part of SR 3. For sure it will be part of 16.10 P2.
In the meantime you can fix it by replacing that .targets file following the steps that I mentioned above: https://github.com/xamarin/xamarin-macios/issues/10109#issuecomment-797062345

**Comment by codeabilitynet** [comment]
@mauroa, FYI, the issue 'error MSB4044: The "GetMinimumOSVersion" task...' persists after the replacement of the Xamarin.Messaging.Build.targets file by the one provided in your .zip.

**Comment by mauroa** [comment]
@codeabilitynet it shouldn't. Please attach a .binlog when you reproduce the error. If it's building with command line, just append `/bl` at the end of the build command. If it's with VS, please follow the next steps:

1. Install the [Project System Tools](https://marketplace.visualstudio.com/items?itemName=VisualStudioProductTeam.ProjectSystemTools) extension
2. In Visual Studio, choose the View > Other Windows > Build Logging menu item.
3. Click on the "play" button.
4. Reproduce the issue
5. Select all entries in the build logging tool window (filtering by Build in the dropdown list)
6. Right-click them, Save Logs

Thanks

**Comment by codeabilitynet** [comment]
@mauroa, 
Please find attached a .zip containing the .binlog file resulting for the MSBuild command run by our CI/CD pipeline on one of our iOS projects. Other iOS projects are similarly affected by the issue.  
[binarylogfilename.zip](https://github.com/xamarin/xamarin-macios/files/6191343/binarylogfilename.zip)

**Comment by codeabilitynet** [comment]
@mauroa, I sent you the .binlog file.

**Comment by codeabilitynet** [comment]
@mauroa, I have been able to reproduce the error on our CI pipeline even with a the default Xamarin.Forms project created by VS.  We are using version 16.9.2 and I had applied the modified Xamarin.Messaging.Build.targets file in the \MSBuild\Xamarin\iOS\ directory. Please find attached the corresponding .log, from which I edited sensitive data. 
[CI.XSB-1.0.0-ios-deploy.log](https://github.com/xamarin/xamarin-macios/files/6228316/CI.XSB-1.0.0-ios-deploy.log)
If usefull, we can send you by email the [corresponding].binlog file.

**Comment by Barokai** [comment]
> I've verified the file that you attached and I can confirm that the fix is not in there, so it didn't make it for SR 2. I'll follow up on this and ensure it's part of SR 3. For sure it will be part of 16.10 P2.
> In the meantime you can fix it by replacing that .targets file following the steps that I mentioned above: [#10109 (comment)](https://github.com/xamarin/xamarin-macios/issues/10109#issuecomment-797062345)

hello,
sadly this isn't fixed with SR3, but i managed to fix it via the provided target file in the zip from the linked comment.

**Comment by cschwarz** [comment]
Also not fixed with 16.9.4, still getting error MSB4044: GetMinimumOSVersion task and the attached Xamarin.Messaging.Build.targets doesn't fix it for me.

**Comment by Delphinidae84** [comment]
I wonder why this issue is still closed, after one month of feedback...

**Comment by mauroa** [comment]
@codeabilitynet Hi. According to your log, you are executing a build with these parameters: `/t:Clean /t:Build`. Both Clean and Build targets needs to connect to the Mac (and disconnect when they finishes). By design, MSBuild only runs a specific target once in the same MSBuild run, so what's happening in this case is that the Clean target connects, then disconnects (because it doesn't now that a Build is coming), and when the Build starts It can't execute the connect target again but the connection is already closed.
I suggest you to change `/t:Clean /t:Build` with `/t:Rebuild`, which will cause our build system to behave different and know that it shouldn't disconnect after the Clean because a Build is coming. Please try this and let me know how it goes. Additionally you can share your build output again.

For @cschwarz and @Delphinidae84, could you please share the MSBuild command that you are using and also the .binlog of the build? See https://github.com/xamarin/xamarin-macios/issues/10109#issuecomment-804880358

Thanks

**Comment by jbrunetti-asi** [comment]
@mauroa & @emaf, this issue is occurring on VS 16.9.4 as well. Same [issue](https://github.com/xamarin/xamarin-macios/issues/10109#issuecomment-821059709) as @cschwarz:

> Also not fixed with 16.9.4, still getting error MSB4044: GetMinimumOSVersion task and the attached Xamarin.Messaging.Build.targets doesn't fix it for me.

What timeframe are we looking at to get this resolved in a stable build of Visual Studio? @mauroa, you had [mentioned](https://github.com/xamarin/xamarin-macios/issues/10109#issuecomment-804128124) this would be in SR3 but we're still breaking on SR4, right?

> I've verified the file that you attached and I can confirm that the fix is not in there, so it didn't make it for SR 2. I'll follow up on this and ensure it's part of SR 3. For sure it will be part of 16.10 P2.
> In the meantime you can fix it by replacing that .targets file following the steps that I mentioned above: [#10109 (comment)](https://github.com/xamarin/xamarin-macios/issues/10109#issuecomment-797062345)

I was not planning on commenting here; but this has been open for quite some time now (since November); and neither of the ([one](https://github.com/xamarin/xamarin-macios/issues/10109#issuecomment-729671062), [two](https://github.com/xamarin/xamarin-macios/issues/10109#issuecomment-797062345)) workarounds mentioned here have worked for my team. Thanks.

**Comment by nschoenberg** [comment]
@jb-asi I Think @mauroa has been super kind and supportive in this thread, trying to help. So maybe try the suggestions, I was able to solve it.

**Comment by nageshshugar** [comment]
> Hi everyone,
> This issue has been fixed for good and will be available on the next VS 16.9 Service Release and the next VS 16.10 Preview 2. However, if you are working on a 16.9 installation, I'm attaching a file that you need to override in the VS installation folder as a workaround to get unblocked until we release the fixed version. So, to apply the workaround please:
> 
> * Go to: `<VS Install path>\MSBuild\Xamarin\iOS`
> * Unzip the attached file (`Xamarin.Messaging.Build.targets.zip`) and paste the .targets file in the above folder. That should replace the existing file with the new one to fix the issue
> * Try to reproduce the issue. It should work now
> 
> Please let me know if that unblocks you.
> Thanks.
> [Xamarin.Messaging.Build.targets.zip](https://github.com/xamarin/xamarin-macios/files/6130076/Xamarin.Messaging.Build.targets.zip)

@taenito Thank you very much. Issue got resolved to me after replacing this file.

**Comment by tossingc** [comment]
We worked around this by **running the build *twice in-a-row***. *(continue on error - **ignore the failed first build**)*

The first time always fails with the error in the bug title, but it must somehow sets things up so that the second build has what it needs to succeed. The IPA is generated, etc.

IMPORTANT: DO **NOT** CLEAN BETWEEN THE TWO BUILDS

----------

*NOTE: As this solution involves an expected failed partial first build, it does add to the build time, but it was only about a minute or two longer per build, and at least it works reliably.*

----------

Hope this is helpful to someone! 🤗

https://stackoverflow.com/questions/64876303/visual-studio-update-16-8-1-lead-to-ci-build-failures/67409349#67409349
https://stackoverflow.com/questions/66051101/the-getminimumosversion-task-was-not-given-a-value-for-the-required-parameter/67409309#67409309
